### ISSUE TITLE
About torch.inverse and NaN

### ISSUE THREAD

**Issue by SongJgit** [anubhabghosh/danse_jrnl#2]
In real applications (not simulated data)， sometimes matrices are irreversible in ``compute_logpdf_Gaussian``. Is there a possible solution?
```bash
torch._C._LinAlgError: linalg.inv: (Batch element 80): The diagonal element 4 is zero, the inversion could not be completed because the input matrix is singular.
```
 For example, during supervised training, loss uses a covariance matrix that looks like this
![image](https://github.com/anubhabghosh/danse_jrnl/assets/30189040/1604cfc8-a5f6-44f2-9cf4-b4e32fc6fc4b)

After our tests, this matrix is indeed irreversible.
![image](https://github.com/anubhabghosh/danse_jrnl/assets/30189040/41d659a8-0ea7-4a99-968d-f5c256871bbd)
Are there any good solutions or approximate solutions to this problem?

As well as, the ``NaN`` problem. In fact, we also tested KalmanNet and the same problem exists.
We believe you have found the same problem as we see NaN checking in the comments of the KalmanNet related code.
Many times the loss returns nan, although compute_predictationu is not a problem.
![image](https://github.com/anubhabghosh/danse_jrnl/assets/30189040/38a025e3-eab1-42f1-80b1-1c5593dddcd4)

**Comment by anubhabghosh** [comment]
I actually worked mostly with diagonal covariances in my problem, so didn't encounter this.

One solution I can think of would be to change the code a bit to work with the precision matrix perhaps instead of the covariance. Maybe, that would help to get rid of the inversion problem. Or if you are working with general covariance matrices, maybe use the RNN to model the cholesky factor of the matrix, i.e. thinking $\mathbf{L} = \pmb{\Sigma}\pmb{\Sigma}^{\top} + \epsilon \mathbf{I}$, where $\epsilon$ is a small positive constant.

**Comment by BarKetPlace** [comment]
@SongJgit there is a short discussion about this in the pytorch documentation as well: https://pytorch.org/docs/stable/distributions.html#multivariatenormal

**Comment by SongJgit** [comment]
@anubhabghosh Thank you for your advice. I've found out what the problem is.
In fact my model is also learning a diagonal covariance, and the reason for this problem is that I was given a wrong C_w, which apparently breaks the premise of diagonal covariance.
If I want the model to do 1-step in, 1-step out during inference, like a normal kalman filter, or n-step in , 1-step out, do you have any good suggestions?
1. $k_{0:n-1}$ in , $k_{n}$ out;  $k_{1:n}$ in , $k_{n+1}$ out？

**Comment by SongJgit** [comment]
I'm sorry to trouble the author.
I have one more question.
If my H is a non-linear like
![image](https://github.com/anubhabghosh/danse_jrnl/assets/30189040/8c901a81-1a38-4b40-9189-ebdd066e72b8)
How do I bring this into the calculation?
like 
![image](https://github.com/anubhabghosh/danse_jrnl/assets/30189040/013a0892-1160-4a08-bef0-4080777b7e17)

**Comment by anubhabghosh** [comment]
The $\mathbf{h}_t$ function looks differentiable to me, you can try linearization similar to 1st-order EKF? Basically expand $\mathbf{h}_t$ with a Taylor-series

**Comment by SongJgit** [comment]
> The ht function looks differentiable to me, you can try linearization similar to 1st-order EKF? Basically expand ht with a Taylor-series

thanks for your advice.
I was wondering if there is another way, after all, jacobian calculations are prone to problems (even with torch.autograd.grad()) as I have encountered with split-kalmannet. I was wondering if I could learn a distribution in the same way as ukf.

**Comment by anubhabghosh** [comment]
Possibly it can be done in the same way as UKF by propagating sigma points as well.

**Comment by anubhabghosh** [comment]
Possibly it can be done in the same way as UKF by propagating sigma points as well. But I guess the issue for the inverse has been resolved, so I'll close the issue now.
